### PR TITLE
StringSegment improvements

### PR DESCRIPTION
**Pull request by javiercn**
* Added a contructor that takes an existing string to allow
representing a string as a StringSegment easily.
* Fix a small bug in SubString and SubSegment.
* Implement IndexOf for StringSegment.
* Optimize Value to return the Buffer when the StringSegment
represents the whole string in the Buffer.

### CODE REVIEW

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 1]
```diff
@@ -6,8 +6,32 @@
...
+    /// </summary>
     public struct StringSegment : IEquatable<StringSegment>, IEquatable<string>
     {
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="buffer">
+        /// The <see cref="string"/>that acts as a buffer for this <see cref="StringSegment"/>.
+        /// In this case, the <see cref="StringSegment"/> represents the whole original <see cref="string"/>
+        /// or its absence in case the parameter is <code>null</code>.
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Can we make this less verbose?

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 2]
```diff
@@ -30,22 +54,61 @@ public StringSegment(string buffer, int offset, int length)
...
+                {
+                    // Typically, the original string was wrapped in a StringSegment in order to make it fit
+                    // some method or member signature, so we just unwrap it.
+                    if (Offset == 0 && Buffer.Length == Length)
+                    {
+                        return Buffer;
+                    }
+                    else
+                    {
+                        return Buffer.Substring(Offset, Length);
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: I believe substring already does this optimization. Measure before you go writing extra code.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 1 above)
Comment (javiercn) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Can you provide a suggestion? I want to remark that this constructor is for creating a StringSegment from the given string and that null buffer is the "equivalent" of null string in this case.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 2 above)
Comment (javiercn) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Yes it does, it was simple enough, I'll change the code, I'll keep the test.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: I'd do something like this. These summaries need to be relatively short. You really **can't** effectively convey lots of info here.

```
The original <see cref="string"/> used as buffer.  The <see cref="StringSegment"/> includes the whole <see cref="string"/>.
```
I'd leave out the part about null.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 3]
```diff
@@ -6,8 +6,32 @@
...
+            Buffer = buffer;
+            Offset = 0;
+            Length = buffer != null ? buffer.Length : 0;
+        }
+
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="buffer">The original <see cref="string"/> used as buffer.</param>
+        /// <param name="offset">The offset within the <paramref name="buffer"/> where the segment starts.</param>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: `The offset of the segment within the <paramref name="buffer"/>.`

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 4]
```diff
@@ -30,22 +54,61 @@ public StringSegment(string buffer, int offset, int length)
...
 
+        /// <summary>
+        /// The offset within the Buffer at wich the segment starts.
+        /// </summary>
         public int Offset { get; }
 
+        /// <summary>
+        /// The length of the segment starting from the offset defined by the segment.
+        /// </summary>
         public int Length { get; }
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: You can clean up all of the docs for these properties to be less verbose as well. It's not saying anything really important.

Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Also add the "Gets or sets" boilerplate

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 5]
```diff
@@ -30,22 +54,61 @@ public StringSegment(string buffer, int offset, int length)
...
 
+        /// <summary>
+        /// The <see cref="string"/> represented by this segment.
+        /// </summary>
         public string Value
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: `Gets the value of this segment as a <see cref="string"/>.`

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 6]
```diff
@@ -30,22 +54,61 @@ public StringSegment(string buffer, int offset, int length)
...
         public bool HasValue
         {
             get { return Buffer != null; }
         }
 
+        /// <summary>
+        /// Checks if the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns><code>true</code> if the specified object is equal to the current object; otherwise, <code>false</code>.</returns>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: inheritdoc

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 7]
```diff
@@ -56,11 +119,22 @@ public override bool Equals(object obj)
...
             return Equals(other, StringComparison.Ordinal);
         }
 
+        /// <summary>
+        /// Checks if the specified <see cref="StringSegment"/> is equal to the current <see cref="StringSegment"/>.
+        /// </summary>
+        /// <param name="other">The <see cref="StringSegment"/> to compare with the current <see cref="StringSegment"/>.</param>
+        /// <param name="comparisonType">One of the enumeration values that specifies the rules to use in the comparison.</param>
+        /// <returns><code>true</code> if the specified <see cref="StringSegment"/> is equal to the current <see cref="StringSegment"/>; otherwise, <code>false</code>.</returns>
         public bool Equals(StringSegment other, StringComparison comparisonType)
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Copy docs from IEquatable<T>

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 8]
```diff
@@ -93,6 +178,10 @@ public bool Equals(string text, StringComparison comparisonType)
             return string.Compare(Buffer, Offset, text, 0, textLength, comparisonType) == 0;
         }
 
+        /// <summary>
+        /// Returns the hash code for this <see cref="StringSegment"/>.
+        /// </summary>
+        /// <returns>A 32-bit signed integer hash code.</returns>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: inheritdoc

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 9]
```diff
@@ -102,23 +191,41 @@ public override int GetHashCode()
             return hash;
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: This hashcode allocates. This is really unfortunate. We should make sure this isn't showing up in our profiles.

**Comment by rynowak** [comment]
Looks good after giving the docs a bit more love :watch:

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 9 above)
Comment (javiercn) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: How does this allocate? HashCodeCombiner is a struct. Is it on any of the calls to Add? when it internally calls s.GetHashCode() or similar?

Comment (javiercn) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: @rynowak How does this allocate? HashCodeCombiner is a struct, is it one of the Add methods that allocates internally?

Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: `.Value` allocates

**Comment by javiercn** [comment]
@rynowak updated

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**